zkevm-circuits/src/rlp_circuit_fsm.rs

```diff
@@ -1405,12 +1497,363 @@ impl<F: Field> RlpCircuitConfig<F> {
+                        meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::prev()) * challenges.keccak_input(),
+                        meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::cur()),
+                    );
+                    cb.require_equal(
+                        "stack_acc accumulates remaining bytes from previous depth level",
+                        meta.query_advice(rlp_decoding_table.stack_acc, Rotation::prev())
+                                + (meta.query_advice(rlp_decoding_table.value, Rotation::prev())
+                                    - meta.query_advice(rlp_decoding_table.value, Rotation::cur())
+                                    // 1 byte needs to be subtracted here as the step preceding the PUSH reads one byte from the lower depth
+                                    // This byte is already consumed and won't be added to the accumulator
+                                    - 1.expr())
+                                * meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::prev()),
+                        meta.query_advice(rlp_decoding_table.stack_acc, Rotation::cur()),
+                    );
+                },
+            );
+
+            // Stack Pop
+            cb.condition(
+                meta.query_advice(rlp_decoding_table.is_stack_pop, Rotation::cur()),
+                |cb| {
+                    cb.require_equal(
+                        "POP stack operation decreases depth",
+                        meta.query_advice(rlp_decoding_table.depth, Rotation::prev()),
+                        meta.query_advice(rlp_decoding_table.depth, Rotation::cur()) + 1.expr(),
+                    );
+                    cb.require_zero(
+                        "POP can only happen if there's no more bytes to decode on the higher depth",
+                        meta.query_advice(rlp_decoding_table.value, Rotation::prev()),
+                    );
+                    cb.require_equal(
+                        "stack_acc accumulates lower depth level at exponent -1 of challenge",
+                        meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::prev()),
+                        meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::cur()) * challenges.keccak_input(),
+                    );
+                    cb.require_equal(
+                        "stack_acc releases remaining bytes from previous depth level",
+                        meta.query_advice(rlp_decoding_table.stack_acc, Rotation::prev())
+                                - meta.query_advice(rlp_decoding_table.value, Rotation::cur())
+                            * meta.query_advice(rlp_decoding_table.stack_acc_pow_of_rand, Rotation::cur()),
+                        meta.query_advice(rlp_decoding_table.stack_acc, Rotation::cur()),
+                    );
```

Comment:
critical: We know it's always true that `stack_acc_prev = RLC(remaining_bytes[0..(prev.depth)], keccak_rand)`. The constraint here only says `stack_acc_cur + value_cur * stack_acc_pow_of_rand = stack_acc_prev`. Note there are two **free** inputs here (i.e. `stack_acc_cur` and `value_cur`), both these two cells' value can be easily controlled by the "malicious" prover. 

I think this is a inherent soundness issue with your approach.